### PR TITLE
Use the latest cabal spec in the interactive package init

### LINKED ISSUE #8977: Chaging the default cabal spec in cabal init from 3.0 to 3.4
I'd like to suggest that we update the default spec that is suggested to the user upon `cabal init`.

Currently the user is offered 

>  * 5) 3.0   (+ set notation for ==, common stanzas in ifs, more redundant commas, better pkgconfig-depends)

by default, but has to manually input 

>    6) 3.4   (+ sublibraries in 'mixins', optional 'default-language')

in order to enjoy the same things + the additions of 3.4.

### PR DESCRIPTION
**Pull request by Kleidukos**
See #8977

This PR updates the version of the Cabal spec that is suggested by default to users when initialising a project interactively. We now use the latest version (3.4).

### QA Notes

Initialising a project with `cabal init` should prompt the user to use the 3.4 spec by default instead of the 3.0. 

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [x] Include [manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) if your PR relates to cabal-install.

Bonus points for added automated tests!

### CODE REVIEW

**Comment by andreabedini** [comment]
Just a nitpick, `cabal init` initialises a package, not a project :relaxed:

**Comment by ulysses4ever** [comment]
It'd still be better to default to the current version of cabal-install imo. Also, it's weird that interactive and non interactive don't do the same.

**Comment by ulysses4ever** [comment]
One reason why I think it's better to explicitly set `latest` is because one day 3.4 will no longer be the latest and then we'll have to decide what to do with the version again (e.g. when to bump it). The whole point of setting it to `latest` always (as discussed in https://github.com/haskell/cabal/issues/8616) is that it allows us to avoid this unnecessary decision process.

**Comment by ulysses4ever** [comment]
@gbaz suggests that we lag one version behind the latest. @fgaz notices that as soon as Cabal spec version is an `Enum` and `Bounded`, it should be as simple as to change the default to `pred maxBound` (instead of the current version, which hard-codes `3.4` or whatever it hard-codes).

This ideally balances my desire to forget about maintenance of this constant forever and some caution about `latest` being too new for some users.

**Comment by ulysses4ever** [comment]
Also, let's make sure that -i and -n default to the same thing.

**Comment by gbaz** [comment]
The broader issue I note is that we haven't updated the `init` parser to even display or parse or let people choose cabal spec versions 3.6 or 3.8.

In general it seems that when we bump the spec, we forget to also fix the `init` system -- maybe there's a way to drive `init` directly off the datatype so this is less possible?

**Comment by Mikolaj** [comment]
@Kleidukos: what's the best course of action with this one? Open new tickets for all the related proposals and merge this one as is or re-evaluate and perhaps fit more into a single PR (in which case, please mark this a draft and remove the needs-review label)?

File: cabal-install/src/Distribution/Client/Init/Interactive/Command.hs [hunk 1]
```diff
@@ -306,7 +306,7 @@ cabalVersionPrompt flags = getCabalVersion flags $ do
     parseCabalVersion "2.4" = CabalSpecV2_4
     parseCabalVersion "3.0" = CabalSpecV3_0
     parseCabalVersion "3.4" = CabalSpecV3_4
-    parseCabalVersion _ = defaultCabalVersion -- 2.4
+    parseCabalVersion _ = defaultCabalVersion -- 3.4
```
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/Init/Interactive/Command.hs:309]: This comment looks like it's doomed to fall behind. Why don't we kill it, and anyone should be able to look up the definition of `defautCabalVersion` themselves?

**Review by ulysses4ever — APPROVED** [review]
@Kleidukos any news here? I'd be happier if `defaultCabalVersion` wasn't hardcoded as it's now and was simply `pred maxBound` (or `pred $ pred maxBound`). But the current version is an obvious improvement, I think.

**Comment by ulysses4ever** [comment]
@Mergify rebase

**Comment by ulysses4ever** [comment]
@mergify refresh

**Review by geekosaur — APPROVED** [review]

**Comment by ulysses4ever** [comment]
On a cabal meeting, we discussed that a more flexible policy may be needed to ensure smooth coordination with HLS, see https://github.com/haskell/cabal/issues/10457

**Pull request closed without merging**